src/rpc/rawtransaction.cpp

```diff
@@ -923,59 +928,72 @@ static RPCHelpMan testmempoolaccept()
         UniValueType(), // VNUM or VSTR, checked inside AmountFromValue()
     });
 
-    if (request.params[0].get_array().size() != 1) {
-        throw JSONRPCError(RPC_INVALID_PARAMETER, "Array must contain exactly one raw transaction for now");
+    if (request.params[0].get_array().size() > MAX_PACKAGE_COUNT) {
+        throw JSONRPCError(RPC_INVALID_PARAMETER, "Array cannot contain more than " + ToString(MAX_PACKAGE_COUNT) + " transactions.");
     }
 
-    CMutableTransaction mtx;
-    if (!DecodeHexTx(mtx, request.params[0].get_array()[0].get_str())) {
-        throw JSONRPCError(RPC_DESERIALIZATION_ERROR, "TX decode failed. Make sure the tx has at least one input.");
-    }
-    CTransactionRef tx(MakeTransactionRef(std::move(mtx)));
-
+    const UniValue raw_transactions = request.params[0].get_array();
     const CFeeRate max_raw_tx_fee_rate = request.params[1].isNull() ?
                                              DEFAULT_MAX_RAW_TX_FEE_RATE :
                                              CFeeRate(AmountFromValue(request.params[1]));
 
     CTxMemPool& mempool = EnsureMemPool(request.context);
-    int64_t virtual_size = GetVirtualTransactionSize(*tx);
-    CAmount max_raw_tx_fee = max_raw_tx_fee_rate.GetFee(virtual_size);
+    CChainState& chainstate = EnsureChainman(request.context).ActiveChainstate();
```

Comment:
Why? We can't do anything in this function if there isn't a mempool, so why not exit early if it doesn't exist?

Reply:
The pattern that we do in every other RPC that handles raw transactions (and most of the ones that take user input that needs to be parsed) is that the parsing/decoding is done first before acquiring anything internal needed for processing. After all, we can't do anything in this function if the transaction is malformed, so why not exit early if it can't be decoded?

I also think that it would make the diff slightly easier to read since that is what the function did originally.

Reply:
I see these `Assume` functions as essentially asserting entry conditions for the function. For rpc methods that require a component to exist in order to function at all (such as `testmempoolaccept`), then I think assume/exit at the top of the function makes most sense. Doing that has benefits:

- avoids unexpected side-effects
- avoid unnecessary lock-taking or resource usage
- explicitly states assumptions up front

There are some rpc methods where components are optional or only required based on particular inputs (eg ` getrawtransaction` might not require a mempool if a blockhash is passed). For those, the `Assume()` function can be further down in the function, but again should be prominently displayed at the top of a code block to document assumptions.